command/agent/agent.go

```diff
@@ -185,56 +164,31 @@ func (c *cmd) run(args []string) int {
 		return 1
 	}
 
-	config := c.readConfig()
-	if config == nil {
-		return 1
-	}
+	logGate := logging.GatedWriter{Writer: &cli.UiWriter{Ui: c.UI}}
 
-	// Setup the log outputs
-	logConfig := &logging.Config{
-		LogLevel:          config.LogLevel,
-		LogJSON:           config.LogJSON,
-		Name:              logging.Agent,
-		EnableSyslog:      config.EnableSyslog,
-		SyslogFacility:    config.SyslogFacility,
-		LogFilePath:       config.LogFile,
-		LogRotateDuration: config.LogRotateDuration,
-		LogRotateBytes:    config.LogRotateBytes,
-		LogRotateMaxFiles: config.LogRotateMaxFiles,
+	agentOptions := []agent.AgentOption{
+		agent.WithBuilderOpts(c.flagArgs),
+		agent.WithCLI(c.UI),
+		agent.WithLogWriter(&logGate),
 	}
-	logger, logGate, logOutput, ok := logging.Setup(logConfig, c.UI)
-	if !ok {
+
+	agent, err := agent.New(agentOptions...)
+	if err != nil {
+		c.UI.Error(err.Error())
 		return 1
 	}
 
-	c.logger = logger
+	config := agent.GetConfig()
+	c.logger = agent.GetLogger()
 
 	//Setup gate to check if we should output CLI information
 	cli := GatedUi{
 		JSONoutput: config.LogJSON,
 		ui:         c.UI,
 	}
 
-	// Setup gRPC logger to use the same output/filtering
-	grpclog.SetLoggerV2(logging.NewGRPCLogger(logConfig, c.logger))
-
-	memSink, err := lib.InitTelemetry(config.Telemetry)
-	if err != nil {
-		c.logger.Error(err.Error())
-		logGate.Flush()
-		return 1
-	}
-
 	// Create the agent
 	cli.output("Starting Consul agent...")
```

Comment:
This log output should go above `agent.New`

Reply:
It cannot go there any longer.

Before the cli code here parsed the config so it already knew prior to creating the agent whether it was doing JSON logging or not. Now it doesn't know that until after `agent.New` and we still want to prevent unstructured logs from being emitted when configured to do JSON logging.

Reply:
However the output was misleading before. It had nothing to do with starting the agent but rather building an object and it would call agent.Start() later on.